### PR TITLE
ANP Re-Conditional Gate: fix OOD-Reynolds regression via adaptive cross-attention

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The merged ANP decoder (PR #2379) showed extraordinary improvements on 3/4 metrics (p_in -70%, p_oodc -48%, p_tan -59%) but **regressed on p_re +16%** (OOD Reynolds generalization).

**Root cause:** The ANP cross-attention unconditionally aggregates fore-foil context for aft-foil predictions. At OOD Reynolds numbers, the fore-foil backbone representations are less reliable (the model hasn't seen this Re distribution during training), so blindly trusting cross-foil attention can propagate fore-foil errors to the aft-foil predictions.

**Fix:** Add a learned, differentiable Re-proximity gate that smoothly scales the cross-attention contribution based on how far the sample's Reynolds number is from the training distribution. In-distribution Re: gate ≈ 1.0 (full cross-attention, identical to current ANP). OOD-Re: gate → smaller, reducing cross-foil interference and falling back toward the backbone residual.

This is a minimal change: 2-layer MLP (16 hidden units) added to the ANP decoder.

## Instructions

Add `--anp_re_gate` flag to `train.py`. When active alongside `--anp_srf`, modifies the ANP cross-attention output for tandem samples.

### Step 1 — Extract Re from batch

Re is in the 24-dim input feature tensor `x`. Check `data/prepare_multi.py` (read-only) to find the Re channel index. Alternatively, if the batch has a flow params dict, extract Re directly from there.

### Step 2 — Add gate network to ANPDecoder `__init__`

```python
if re_gate:
    self.re_gate_mlp = nn.Sequential(
        nn.Linear(1, 16),
        nn.Tanh(),
        nn.Linear(16, 1),
        nn.Sigmoid()
    )
    # Initialize bias to +5 so gate starts near 1.0 (preserves ANP behavior at init)
    nn.init.constant_(self.re_gate_mlp[2].bias, 5.0)
```

### Step 3 — Compute gate in forward pass

```python
if self.re_gate:
    # log_Re_norm: shape [B, 1]
    log_re = torch.log(re_values.clamp(min=1000.0)).unsqueeze(-1)  # [B, 1]
    # Approx training stats: Re range ~40k-200k → log mean ≈ 11.3, log std ≈ 0.5
    log_re_norm = (log_re - 11.3) / 0.5
    gate = self.re_gate_mlp(log_re_norm)  # [B, 1]
    gate = gate.unsqueeze(1)  # [B, 1, 1] for broadcasting over surface nodes
```

### Step 4 — Apply gate to cross-attention output (aft-foil nodes only)

After computing `cross_attn_out` (output from cross-attention), for tandem samples:

```python
# For aft-foil nodes that received cross-foil context:
anp_out_aft = gate * cross_attn_out_aft + (1 - gate) * backbone_residual_aft
# For fore-foil nodes (self-attention only): unchanged
anp_out_fore = cross_attn_out_fore
```

The backbone residual is already available from the existing `+ hidden` residual in the ANP. Use that as the fallback.

**Fallback**: If extracting Re cleanly is difficult, implement `--anp_re_context` instead: concatenate log(Re) as an extra feature to the ANP context encoder input (alongside backbone hidden state + cp_panel + coords). Let the attention heads learn Re-dependent weighting naturally. Try `--anp_re_gate` first; use `--anp_re_context` if implementation is simpler.

### Training Command (run both seeds)

```bash
cd cfd_tandemfoil && python train.py \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --anp_srf --anp_re_gate \
  --wandb_group anp-re-conditional \
  --seed 42
```

### What to Watch

- **p_re** is the PRIMARY target — recover toward or below 6.229 (old baseline) while keeping p_tan < 10.825
- **p_tan/p_in/p_oodc** must NOT regress significantly — gate initializes to 1.0 so in-distribution should be preserved
- Log `gate.mean()` to W&B to verify the gate is learning (should be 1.0 for in-dist, lower for OOD)
- If gate collapses (always 0 or always 1), reduce init bias from +5 to +2

## Baseline (PR #2379 ANP — now current)

| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **3.561** | < 3.561 |
| **p_oodc** | **3.847** | < 3.847 |
| **p_tan** | **10.825** | < 10.825 |
| **p_re** | **7.232** | **< 6.229** ← priority: recover to old baseline |

W&B runs: `metqdxaq` (seed 42), `jvfvrs4u` (seed 73)

Reproduce baseline:
```
cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 --anp_srf
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2390 because ADVISOR: Closing — all 3 debug-re-gate W&B runs crashed at step 0 with no metrics (kjg24eex, k8kh0sll, n9et63t0). GPU has been idle 7+ hours. The Re-conditional gate idea is conceptually sound but the implementation has a persistent crash bug. Reassigning you to a fresh P0 experiment (BL Adaptive Node Weighting) — see new PR..
